### ISSUE TITLE
kubelet can't see images loaded into containerd

### ISSUE THREAD

**Issue by dhiltgen** [containerd/cri#760]
If you use `ctr` (or the containerd API) to load images into the `k8s.io` namespace, those images are not visible to the kubelet.  This is due to an extra layer of metadata/caching taking place in the CRI layer.  

This extra layer adds complexity and inconsistency into the stack, which could lead to problems in production environments.  Tools developed against containerd will get an inconsistent view of the images compared to the kubelet.  Operators may take action with incomplete/inconsistent information, and cause failures in their environment.

We should try to find a way to have a single source of truth for the images loaded within the namespace.

**Comment by Random-Liu** [comment]
Questions:
* How are we going to integrate containerd image management with moby? There will be similar issue. Do we expect image pulled by `ctr` in `moby.io` seen by Docker?
* Do we expect container created by `ctr` been seen by CRI? Does Docker see container created by `ctr` in `moby.io`?

As for how to solve this problem, we need to change both `ctr` and `cri`:
1. For `ctr`, we should change `ctr pull` to behave like `docker pull`, which indicates `ctr pull` should:
  a. Write image id into image metadata store, so as to index image by id (required);
  b. Store repo digests and repo tags as image labels (required);
  c. Write other known image tags into metadata store, such as `digest` and `tag`, to make it easy to reference (prefered, we need to traverse image store to find a image tag/digest);
  d. Write image metadata after the image is unpacked (prefered, without this the first container start could be super slow and potentially timeout).
2. For `cri`, we need to get rid of the image cache, and always read containerd image metadata store.

As long as 1 is done, 2 should be easy.

**Comment by abhi** [comment]
@Random-Liu , @crosbymichael had a suggestion on using filters. containerd can store in the metadata with filter labels and we can use filter selectors in cri to get the image from the metadata store instead of storing all the different look up keys.

**Comment by dmcgowan** [comment]
I would like to understand which requests some of these are being used for. Seems pretty straightforward.

> a. Write image id into image metadata store, so as to index image by id (required);

Storing the image is is simple. How is the lookup done though? As @abhi mentioned just adding to a label and filtering could achieve this.

> b. Store repo digests and repo tags as image labels (required);

Repo digest could be reconstructed by the client from the current image object. Curious how the other tags are used.

> c. Write other known image tags into metadata store, such as digest and tag, to make it easy to reference (prefered, we need to traverse image store to find a image tag/digest);

Would just having the image id in the labels allow you build this from the list of images without any additional requests?

> d. Write image metadata after the image is unpacked (prefered, without this the first container start could be super slow and potentially timeout).

I agree with this one, actually part of another change I am making right now

**Comment by Random-Liu** [comment]
@abhi @crosbymichael Then by default how should we index an image?

1) If we index the image with image tag (what we are doing in `ctr` today), what happens if the image tag is pointed to a newer image? After pulling the new image, the previous image tag will be overwritten?
2) If we index the image with image id/digest, what happens if the image tag is pointed to a newer image? We may need to update all existing image with that tag, and delete it from their labels.

**Comment by Random-Liu** [comment]
This won't be a backward compatible change. We need to decide:
1) Either suggest user to delete all images before upgrade;
2) Or handle old image metadata and convert them to new format during restart.

**Comment by dmcgowan** [comment]
We have a database migration framework already built in for metadata at https://github.com/containerd/containerd/blob/master/metadata/migrations.go . We only need that if we change format and shouldn't be needed by plugins. If it is just label changes, that can be accomplished on plugin startup.

**Comment by Random-Liu** [comment]
> We only need that if we change format and shouldn't be needed by plugins. If it is just label changes, that can be accomplished on plugin startup.

Yeah, I mean that. :) During containerd restart, CRI plugin can deal with old image metadata, and generate new ones with labels.

**Comment by Random-Liu** [comment]
Using label filter sounds good to me. But I still have question about what to use as the default image index. https://github.com/containerd/cri/issues/760#issuecomment-386742191 And I'm curious how docker handles the case that image tag points to a different image, e.g. `:latest` image.

**Comment by Random-Liu** [comment]
## One Proposed Solution
How about this? In containerd client, we do this:
1) Let docker resolver return `repodigest`.
2) After pulling down the image, we write both `repodigest` and the initial image ref into metadata store if it is different from `repodigest` (should be `repotag` in this case).
3) Both the image references have image id as a label.
4) When index by `repotag` or `repodigest`, we can get it from image metadata store directly.
5) When index by image id, we can use label filter.
6) When a new image with the same `repotag` is pulled, the original `repotag` will be overwritten, but the `repodigest` is still there. (This is also docker's behavior today)

**Comment by stevvooe** [comment]
@Random-Liu This is a good plan but we may need to index the labels in the future, depending on performance requirements. In general, it has worked and will continue to work for small n.

I'm in favor of option 1, at first glance. We should be able to write a unique reference, such as the manifest digest, into the image store to satisfy these requirements.

**Comment by Random-Liu** [comment]
> This is a good plan but we may need to index the labels in the future, depending on performance requirements. In general, it has worked and will continue to work for small n.

Actually kubelet is more sensitive to container list/inspection performance today. Heavier image operation shouldn't be too much a problem.

But if we can write image ID as well, we can completely get rid of the performance issue. Basically this means that we write all image references into image metadata store, including tag, digest and id. Are we ok with that?

**Comment by lbernail** [comment]
Yes I ran into this yesterday. It's not really a big issue for us because most of the time the image pull is triggered by the kubelet.
However we hit a quota issue on the image repo and were not able to pull images for a while so I tried to fix some issues using manual import.

Another use-case for us would be the ability to preload images when we create our instance images (but for this we can do a `crictl image pull`)
